### PR TITLE
[MRG] Add object detection query sampler

### PR DESCRIPTION
**Pull request by AlexandreAbraham**
Still WIP. Adds a query sampler for object detection using the uncertainty strategies on the bounding box with highest prediction probability.

### CODE REVIEW

File: custom-recipes/object-detection-query-sampler/recipe.py [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+    boxes.set_shape([max_detections, 4])
+    scores.set_shape([max_detections, int(classification.shape[1])])
+    labels.set_shape([max_detections])
+    for o, s in zip(other_, [list(keras.backend.int_shape(o)) for o in other]):
+        o.set_shape([max_detections] + s[1:])
+
+    return [boxes, scores, labels] + other_
+
+
+class MyFilterDetections(FilterDetections):
```
Comment (dsleo) [inline custom-recipes/object-detection-query-sampler/recipe.py:169]: I'm missing how is this different from the original [FilterDetection](https://github.com/fizyr/keras-retinanet/blob/f4017fdfb942ef594bf3f81d846f6e6100fa9e5d/keras_retinanet/layers/filter_detections.py#L115) ?

File: custom-recipes/object-detection-query-sampler/recipe.py [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+        outputs = backend.map_fn(
+            _filter_detections,
+            elems=[boxes, classification, other],
+            dtype=[keras.backend.floatx(), keras.backend.floatx(), 'int32'] + [o.dtype for o in other],
+            parallel_iterations=self.parallel_iterations
+        )
+
+        return outputs
+
+    def compute_output_shape(self, input_shape):
```
Comment (dsleo) [inline custom-recipes/object-detection-query-sampler/recipe.py:209]: I might be missing but this looks exactly like the original [one](https://github.com/fizyr/keras-retinanet/blob/f4017fdfb942ef594bf3f81d846f6e6100fa9e5d/keras_retinanet/layers/filter_detections.py#L184) ?

File: custom-recipes/object-detection-query-sampler/recipe.py [hunk 3]
```diff
@@ -0,0 +1,487 @@
+import os.path as op
```
[the remark this answers is not in the document]
Comment (andreybavt) [reply custom-recipes/object-detection-query-sampler/recipe.py]: not relevant anymore

File: custom-recipes/object-detection-query-sampler/recipe.py [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+                                        configs.get('gpu_allocation', 0.))
+
+batch_size = 3  # int(configs['batch_size'])
+confidence = 0.1  # float(configs['confidence'])
+
+model = get_test_model(weights, len(labels_to_names))
+
+df = pd.DataFrame(columns=['path', 'uncertainty', 'session'])
+df_idx = 0
+session = 1
```
Comment (andreybavt) [inline custom-recipes/object-detection-query-sampler/recipe.py:418]: session doesn't increase

**Comment by andreybavt** [comment]
Do we also need to add 
```
git+https://github.com/fizyr/keras-retinanet.git@0.5.1
opencv-python==3.4.0.12
```

into the requirements for the recipe to work?

File: custom-recipes/object-detection-query-sampler/recipe.py (hunk 4 above)
Comment (AlexandreAbraham) [reply custom-recipes/object-detection-query-sampler/recipe.py]: Outdated

**Comment by AlexandreAbraham** [comment]
Yes, I have added keras-retinanet. opencv should not be required as it is installed by retinanet

File: custom-recipes/object-detection-query-sampler/recipe.py [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+from keras_retinanet.utils.model import freeze as freeze_model
+from keras_retinanet.utils.image import read_image_bgr, preprocess_image, resize_image
+from keras_retinanet import backend
+from keras_retinanet.layers import FilterDetections
+from lal import utils
+from cardinal import uncertainty
+
+
+
+def prettify_error(s):
```
Comment (andreybavt) [inline custom-recipes/object-detection-query-sampler/recipe.py:28]: This code is duplicated 3 times: in 2 recipes and in the helper. Let's maybe put 1 version to utils.py and import it from there?

File: custom-recipes/object-detection-query-sampler/recipe.py [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+    except:
+        raise ImportError(prettify_error(
+            'ML assisted labeling object detection relies on the DSS object detection '
+            'plugin. The plugin could not be found on this instance. Please insteall it '
+            'to use this recipe. More information: https://www.dataiku.com/product/plugins/object-detection/'
+        ))
+
+
+from retinanet_model import get_model
+from dfgenerator import DfGenerator
```
Comment (andreybavt) [inline custom-recipes/object-detection-query-sampler/recipe.py:51]: not used?

File: detectRecipeController.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
+var app = angular.module('detectionRecipe.detect', []);
```
Comment (andreybavt) [inline detectRecipeController.js:1]: Duplicated file: js/detectRecipeController.js

File: custom-recipes/object-detection-query-sampler/recipe.py [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+import dataiku
+from dataiku.customrecipe import *
+
+import pandas as pd
+import numpy as np
+import tensorflow as tf
+
+from sklearn.model_selection import train_test_split
+from keras.utils import multi_gpu_model
+from keras.models import load_model
```
[the remark this answers is not in the document]
Comment (AlexandreAbraham) [reply custom-recipes/object-detection-query-sampler/recipe.py]: True my vscode does not detect them anymore I will fix that.

File: custom-recipes/object-detection-query-sampler/recipe.py (hunk 5 above)
Comment (AlexandreAbraham) [reply custom-recipes/object-detection-query-sampler/recipe.py]: Fixed.

File: custom-recipes/object-detection-query-sampler/recipe.py (hunk 6 above)
Comment (AlexandreAbraham) [reply custom-recipes/object-detection-query-sampler/recipe.py]: Fixed.

File: detectRecipeController.js (hunk 7 above)
Comment (AlexandreAbraham) [reply detectRecipeController.js:1]: Fixed.

**Pull request merged**